### DISCUSSION TITLE
Mouse And Keyboard Emulation

### DISCUSSION THREAD

**Discussion by Magic4k** [seleniumbase/SeleniumBase#2749] [Q&A] [answered]
How To make a specific and customized mouse and keyboard emulation in my tests ? like move and click ...scrolling...duration of moving ...random mouse emulation ...etc
is it possible in uc mode ?

**Comment by mdmintz** [comment, marked as answer]
Mouse Actions come from `ActionChains`, which is part of raw Selenium:
https://www.selenium.dev/documentation/webdriver/actions_api/mouse/
With SeleniumBase, you can use any raw Selenium method. (`sb.driver`)

There's another section for Keyboard Actions:
https://www.selenium.dev/documentation/webdriver/actions_api/keyboard/

The complete Actions API:
https://www.selenium.dev/documentation/webdriver/actions_api/
That includes a Mouse Wheel API:
https://www.selenium.dev/documentation/webdriver/actions_api/wheel/

For anything not covered above, there's a separate repo called `pyautogui`:
https://github.com/asweigart/pyautogui
(GUI automation used to programmatically control the mouse & keyboard.)

**Reply by bogdanblazhkevych** [reply]
I read in one of your replies to another discussion question that it is not recommended to use WebDriverWait if you want to remain undetected, does this also apply to ActionChains? Specifically, the move_by_offset method.

**Reply by mdmintz** [reply]
That depends on the website. If you can, use https://github.com/asweigart/pyautogui for that kind of thing.

**Reply by bogdanblazhkevych** [reply]
> That depends on the website. If you can, use https://github.com/asweigart/pyautogui for that kind of thing.

This website tracks mouse movement so I'm trying to choose the best possible option.

I've seen this suggested before but I'm looking to do multithreading and I'm not sure that's going to work too well.

Is there a way to do this without getting into expensive cloud solutions?

**Reply by mdmintz** [reply]
You can do multithreading with SeleniumBase: https://github.com/seleniumbase/SeleniumBase/issues/2478#issuecomment-1981699298
It's covered in the YouTube video: https://www.youtube.com/watch?v=2pTpBtaE7SQ

**Reply by bogdanblazhkevych** [reply]
Yes it does, but pyautogui doesn't like it so much, especially when doing parallel actions across multiple threads

**Reply by mdmintz** [reply]
Correct, `pyautogui` isn't good for parallel execution on the same machine. However, I'm the maintainer of `seleniumbase`, not `pyautogui`, so I can't make `pyautogui` better. Use one thread per machine when using `pyautogui`.
